packages/browser-repl/src/lib/runtime-helpers/open-context-runtime.ts

```diff
@@ -0,0 +1,65 @@
+import { Completion } from '../autocompleter/autocompleter';
+import { ServiceProvider } from 'mongosh-service-provider-core';
+import { ShellApiAutocompleter } from '../autocompleter/shell-api-autocompleter';
+import { Interpreter, InterpreterEnvironment, EvaluationResult } from '../interpreter';
+import { Runtime } from '../../components/runtime';
+
+import Mapper from 'mongosh-mapper';
+import ShellApi from 'mongosh-shell-api';
+
+/**
+ * This class is the core implementation for a runtime which is not isolated
+ * from the environment of the presentation layer.
+ *
+ * This means that the interaction between the runtime and the execution context (
+ * service provider, autocompleter, ...), can happen through direct method
+ * calls rather than requiring event emitter bridges or RPC.
+ */
+export class OpenContextRuntime implements Runtime {
+  private serviceProvider: ServiceProvider;
+  private interpreter: Interpreter;
+  private autocompleter: ShellApiAutocompleter;
+
+  constructor(serviceProvider: ServiceProvider, interpreterEnvironment: InterpreterEnvironment) {
+    this.serviceProvider = serviceProvider;
+
+    this.setupEvaluationContext(
+      interpreterEnvironment.getContextObject(),
+      serviceProvider
+    );
+
+    this.interpreter = new Interpreter(interpreterEnvironment);
+  }
+
+  async getCompletions(code: string): Promise<Completion[]> {
```

Comment:
I don't think I quite understand why the completion function has to be in this open context runtime as well? is it because you get the server version first and that can only happen from the service provider?

what if the server version gets added to the `ns`? or just even the redux state so you don't have to set up completions in `runtime`?

Reply:
The whole idea of the `Runtime` is for the top level/react-specific stuff to not have any direct dependency on shell-api / mapper and service provider, since that would not work with WebWorkers or more isolated environments.

I agree 100% we should do differently otherwise. I hope we can sort out If we need that or not, we may drop the runtime object entirely in that case.